### PR TITLE
nixos/nix-daemon: Add enable option

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

Don't worry, it's is true by default. But I think this is important to
have because NixOS indeed shouldn't need Nix at run time when the
installation is not being modified, and now we can verify that.

NixOS images that cannot "self-modify" are a legitamate
use-case that this supports more minimally. One should be able to e.g. do a
sshfs mount and use `nixos-install` to modify them remotely, or just
discard them and build fresh ones if they are run VMs or something.

The next step would be to make generations optional, allowing just
baking `/etc` and friends rather than using activation scripts. But
that's more involved so I'm leaving it out.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lukegb** [comment]
I love the concept of this - I've been toying with this idea myself as well. Being able to build a NixOS image *using* Nix but without Nix in the runtime closure is a super interesting idea.

**Comment by Ericson2314** [comment]
Ugh, i forgot about `nixosTests`. The layering violation leaves me deeply unsettled.

**Comment by tomberek** [comment]
Is the `pkgsFun` portion needed to completely remove nix from the closure?

**Comment by Ericson2314** [comment]
@tomberek Not sure, maybe I'll save that for later.

File: nixos/tests/without-nix.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  pkgs = pkgsFun {
+    overlays = [
+      (self: super: {
+        nix = throw "don't want to use this";
+      })
+    ];
+  };
+in
+
+with import ../lib/testing-python.nix { inherit system pkgs; };
```
Comment (roberth) [inline nixos/tests/without-nix.nix:19]: I don't think `pkgs` makes it all the way into the NixOS config, similar to https://github.com/NixOS/nixpkgs/issues/55798.

If it does, you could prove this important test assumption by adding another attribute in your overlay and adding an assertion about it in the NixOS config.

Comment (sternenseemann) [reply nixos/tests/without-nix.nix:19]: The `pkgs` inherited here is the only one available in `../lib/testing-python.nix`, so why not?

**Pull request merged**